irohad/torii/impl/command_service_impl.cpp

```diff
@@ -150,30 +153,91 @@ namespace iroha {
 
     void CommandServiceImpl::processBatch(
         std::shared_ptr<shared_model::interface::TransactionBatch> batch) {
-      tx_processor_->batchHandle(batch);
+      const auto status_issuer = "ToriiBatchProcessor";
       const auto &txs = batch->transactions();
-      std::for_each(txs.begin(), txs.end(), [this](const auto &tx) {
-        const auto &tx_hash = tx->hash();
-        auto found = cache_->findItem(tx_hash);
-        // StatlessValid status goes only after
-        // EnoughSignaturesCollectedResponse So doesn't skip publishing status
-        // after it
-        if (found
-            and iroha::visit_in_place(
-                    found.value()->get(),
-                    [](const shared_model::interface::
-                           EnoughSignaturesCollectedResponse &) {
-                      return false;
-                    },
-                    [](auto &) { return true; })
-            and tx->quorum() < 2) {
-          log_->warn("Found transaction {} in cache, ignoring", tx_hash.hex());
+
+      bool cache_has_at_least_one_tx{false};
+      bool batch_has_mst_pending_tx{false};
+      std::tie(cache_has_at_least_one_tx, batch_has_mst_pending_tx) =
+          std::accumulate(
+              txs.begin(),
+              txs.end(),
+              std::make_pair<bool, bool>(false, false),
+              [this, &status_issuer](std::pair<bool, bool> lookup_result,
+                                     const auto &tx) {
+                const auto &tx_hash = tx->hash();
+                if (auto found = cache_->findItem(tx_hash)) {
+                  iroha::visit_in_place(
+                      (*found)->get(),
+                      [this, &found, &lookup_result, &status_issuer](
+                          const shared_model::interface::MstPendingResponse &) {
+                        this->pushStatus(status_issuer, *found);
+                        lookup_result.second = true;
+                      },
+                      [this, &tx_hash, &status_issuer](
+                          const shared_model::interface::NotReceivedTxResponse
+                              &) {
+                        this->pushStatus(
+                            status_issuer,
+                            status_factory_->makeStatelessValid(tx_hash));
+                      },
+                      [this, &found, &status_issuer](const auto &status) {
+                        this->pushStatus(status_issuer, *found);
+                      });
+                  lookup_result.first = true;
+                }
+                return lookup_result;
+              });
+
+      if (cache_has_at_least_one_tx and not batch_has_mst_pending_tx) {
+        return;
+      }
+
+      if (not batch_has_mst_pending_tx) {
```

Comment:
Is it possible to pass batch with pending txes which was already processed?

Reply:
fixed